### PR TITLE
Allow the same variable in read and write in USEION

### PR DESCRIPTION
**Pull request by cattabiani**
Fix issue #215

### CODE REVIEW

**Comment by cattabiani** [comment]
please retest

File: src/codegen/codegen_c_visitor.cpp [hunk 1]
```diff
@@ -583,26 +583,34 @@ std::vector<std::string> CodegenCVisitor::ion_read_statements(BlockType type) {
...
-            if (type == BlockType::Ode && ion.is_ionic_conc(var) && state_variable(var)) {
+            if (type == BlockType::Ode && ion.is_ionic_conc(var) && state_variable(var) ||
+                no_doubles.find(var) != no_doubles.end()) {
                 continue;
             }
             auto variable_names = read_ion_variable_name(var);
             auto first = get_variable_name(variable_names.first);
             auto second = get_variable_name(variable_names.second);
             statements.push_back("{} = {};"_format(first, second));
+            no_doubles.emplace(var);
```
Comment (ohm314) [inline src/codegen/codegen_c_visitor.cpp:601]:
```suggestion
            no_doubles.insert(var);
```
You don't need to construct.

File: src/codegen/codegen_c_visitor.cpp [hunk 2]
```diff
@@ -583,26 +583,34 @@ std::vector<std::string> CodegenCVisitor::ion_read_statements(BlockType type) {
         return ion_read_statements_optimized(type);
     }
     std::vector<std::string> statements;
+    std::unordered_set<std::string> no_doubles;  // used to keep track of the variables to not
```
[the remark this answers is not in the document]
Comment (ohm314) [reply src/codegen/codegen_c_visitor.cpp:586]: maybe even name it something like `ion_vars` or something like that?

File: src/codegen/codegen_c_visitor.cpp [hunk 3]
```diff
@@ -875,19 +883,28 @@ std::vector<IndexVariableInfo> CodegenCVisitor::get_int_variables() {
...
-            variables.back().is_constant = true;
-        }
+        std::unordered_set<std::string> no_doubles;  // used to keep track of the variables to not
+                                                     // have doubles between read/write. Same name
+                                                     // variables are allowed as required in issue
+                                                     // #215
         for (const auto& var: ion.writes) {
-            variables.emplace_back(make_symbol("ion_" + var));
+            const std::string name = "ion_" + var;
+            no_doubles.emplace(name);
```
Comment (ohm314) [inline src/codegen/codegen_c_visitor.cpp:892]:
```suggestion
            no_doubles.insert(name);
```

File: azure-pipelines.yml [hunk 4]
```diff
@@ -28,17 +28,35 @@ jobs:
...
       then
         make VERBOSE=1
       fi
+      make install
       env CTEST_OUTPUT_ON_FAILURE=1 make test
     env:
       CMAKE_PKG: 'cmake-3.10.2-Linux-x86_64'
-    displayName: 'Build and Run Tests'
+    displayName: 'Build and Run Unit Tests'
+  - script: |
```
[the remark this answers is not in the document]
Comment (cattabiani) [reply azure-pipelines.yml:43]: maybe for another pr?

File: src/codegen/codegen_c_visitor.cpp (hunk 1 above)
Comment (cattabiani) [reply src/codegen/codegen_c_visitor.cpp:601]: I do not understand. Both insert and emplace construct a copy of the object. I do not see any advantage with insert over emplace

**Comment by pramodk** [comment]
I didn’t get time to review this yet. I would like compare this implementation with neuron/coreneuron  because this is quite a special case.

**Comment by cattabiani** [comment]
please retest

**Comment by cattabiani** [comment]
please retest

**Comment by cattabiani** [comment]
please retest

File: src/codegen/codegen_c_visitor.cpp (hunk 1 above)
Comment (cattabiani) [reply src/codegen/codegen_c_visitor.cpp]: done

**Comment by pramodk** [comment]
Some improvements that I am noticing with the PR:

* New warnings

```
[ 88%] Built target testlexer
/Users/kumbhar/workarena/repos/bbp/incubator/nocmodl/src/codegen/codegen_c_visitor.cpp:592:66: warning: '&&' within '||' [-Wlogical-op-parentheses]
            if (type == BlockType::Ode && ion.is_ionic_conc(var) && state_variable(var) ||
                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~ ~~
/Users/kumbhar/workarena/repos/bbp/incubator/nocmodl/src/codegen/codegen_c_visitor.cpp:592:66: note: place parentheses around the '&&' expression to silence this warning
            if (type == BlockType::Ode && ion.is_ionic_conc(var) && state_variable(var) ||
                                                                 ^
                (                                                                      )
/Users/kumbhar/workarena/repos/bbp/incubator/nocmodl/src/codegen/codegen_c_visitor.cpp:603:66: warning: '&&' within '||' [-Wlogical-op-parentheses]
            if (type == BlockType::Ode && ion.is_ionic_conc(var) && state_variable(var) ||
                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~ ~~
/Users/kumbhar/workarena/repos/bbp/incubator/nocmodl/src/codegen/codegen_c_visitor.cpp:603:66: note: place parentheses around the '&&' expression to silence this warning
            if (type == BlockType::Ode && ion.is_ionic_conc(var) && state_variable(var) ||

```

* Uniform indentation/white spaces missing for https://github.com/BlueBrain/nmodl/pull/346/files#diff-0d42e7f6087abd0e3ac1dc3773b54c7c

**Comment by pramodk** [comment]
I am checking generated code by using  [calcium_dynamics.mod.txt](https://github.com/BlueBrain/nmodl/files/4838037/calcium_dynamics.mod.txt) (just modified example), and generated code is not correct between mod2c of neuron/coreneuron and NMODL.

e.g.

md2c
```cpp
#define _ion_ica        _nt_data[_ppvar[0*_STRIDE]]
#define _ion_cai    _nt_data[_ppvar[1*_STRIDE]]
#define _ion_cao    _nt_data[_ppvar[2*_STRIDE]]
#define _ion_dicadv _nt_data[_ppvar[3*_STRIDE]]
#define _style_ca   _ppvar[4]
```

NMODL

```cpp
        inst->ion_ica = nt->_data;
        inst->ion_dicadv = nt->_data;
        inst->ion_cai = nt->_data;
        inst->ion_cao = nt->_data;
        inst->style_ca = ml->pdata;
```

and as the order is different, they will result into reading/writing wrong values.

@ohm314 : I think you missed my comment https://github.com/BlueBrain/nmodl/pull/346#issuecomment-645790740 :)

**Comment by cattabiani** [comment]
1- I never tested with  --cse codegen --force
2- I did not notice that the order in which the variables was declared was important

**Comment by pramodk** [comment]
> 1- I never tested with --cse codegen --force

Note that `--cse codegen --force` doesn't matter here. What I was trying to see is if mod file provided by Michale is handled correctly in NMODL.

> 2- I did not notice that the order in which the variables was declared was important

Ok. The reason for this is that the in-memory data structure is passed from NEURON to CoreNEURON and variables are in memory at specific offsets. If you change the order, you get different memory segments.

**Pull request merged**